src/fireedge/src/server/routes/api/marketapp/functions.js

```diff
@@ -0,0 +1,301 @@
+    associated === 'false' && paramsCommand.push('--no')
+    tag && paramsCommand.push('--tag', tag)
+    template && paramsCommand.push('--template', template)
+    vmname && paramsCommand.push('--vmname', vmname)
+    user && paramsCommand.push('--user', user)
+    password && paramsCommand.push('--password', password)
+
+    const executedCommand = executeCommand(
+      defaultCommandMarketApp,
+      paramsCommand,
+      prependCommand
+    )
+    const response = executedCommand.success ? ok : internalServerError
+    if (executedCommand.data) {
+      message = executedCommand.data
+    }
+    rtn = httpResponse(response, message)
+  }
+  res.locals.httpCode = rtn
+  next()
+}
+
+/**
+ * Exports the marketplace app to the OpenNebula cloud.
+ *
+ * @param {object} res - http response
+ * @param {Function} next - express stepper
+ * @param {object} params - params of http request
+ * @param {number} params.id - app id
+ * @param {object} userData - user of http request
+ * @param {Function} oneConnection - function of xmlrpc
+ */
+const downloadApp = (
+  res = {},
+  next = defaultEmptyFunction,
+  params = {},
+  userData = {},
+  oneConnection = defaultEmptyFunction
+) => {
+  const { id, token } = params
+  if (!(Number.isInteger(parseInt(id, 10)) && token)) {
+    responseHttp(res, next, httpNotFoundRequest)
+
+    return
+  }
+
+  const userDataFromJWT =
+    validateAuth({
+      headers: { authorization: token },
+    }) || {}
+  const { aud, jti } = userDataFromJWT
+
+  if (!(aud && jti)) {
+    responseHttp(res, next, httpResponse(unauthorized, '', ''))
+
+    return
+  }
+
+  const oneConnect = oneConnection(aud, jti)
+  const callbackNotfound = () => responseHttp(res, next, httpNotFoundRequest)
+  const market = ({ MARKETPLACE_ID, SOURCE }) => {
+    Number.isInteger(parseInt(MARKETPLACE_ID, 10)) &&
+      getMarket({
+        oneConnect,
+        id: MARKETPLACE_ID,
+        success: (MARKET) => {
+          const drvMessage = `<DS_DRIVER_ACTION_DATA>${MARKET}</DS_DRIVER_ACTION_DATA>`
+          const drvMessageBase64 = btoa(drvMessage)
+          const downloadCmd = `DRV_ACTION=${drvMessageBase64}; ${global.paths.DOWNLOADER} ${SOURCE} -`
+          const filename = `one-marketplaceapp-${id}`
+          res.setHeader('Content-Type', 'application/octet-stream')
+          res.setHeader('Cache-Control', 'no-transform')
+          res.setHeader(
+            'Content-Disposition',
+            `attachment; filename=${filename}`
+          )
+          const execChild = exec(
+            downloadCmd,
+            { maxBuffer: 1024 ** 3, encoding: '' },
+            (error) => {
+              error &&
+                writeInLogger(error) &&
+                messageTerminal({
+                  color: 'red',
+                  message: 'error download marketapp: %s',
+                  error,
+                })
+            }
+          )
+          execChild.stdout.pipe(res)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Server-Side Request Forgery (SSRF) vulnerability detected.  Untrusted user input is being used in a network request (in fetch, axios, or http.request)  without proper validation. This can allow attackers to craft requests to internal systems, access metadata services,  or perform internal port scans. To fix this issue, strictly validate and sanitize the input URL  using a library like `node:url` for parsing, enforce an allowlist of trusted domains,  and block access to private IP ranges (e.g., 127.0.0.1, 169.254.169.254).  Learn more: https://owasp.org/www-community/attacks/Server_Side_Request_Forgery